### PR TITLE
HIVE-24624: Repl Load should detect the compatible staging dir

### PR DESCRIPTION
**Pull request by pmadhukar-cloudera**

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java [hunk 1]
```diff
@@ -177,6 +177,7 @@ public int execute() {
             initiateAuthorizationDumpTask();
           }
           DumpMetaData dmd = new DumpMetaData(hiveDumpRoot, conf);
+          dmd.setVersion("V3");
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/repl/ReplDumpTask.java:180]: Add as a constant

File: ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java [hunk 2]
```diff
@@ -359,6 +359,9 @@ private void analyzeReplLoad(ASTNode ast) throws SemanticException {
       if (loadPath != null) {
         DumpMetaData dmd = new DumpMetaData(loadPath, conf);
 
+        if (!dmd.isVersionCompatible()) {
+          throw new SemanticException("Dump version older than V3 is not supported.");
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java:363]: print the dump version also. V3 or 3?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java [hunk 3]
```diff
@@ -359,6 +359,10 @@ private void analyzeReplLoad(ASTNode ast) throws SemanticException {
       if (loadPath != null) {
         DumpMetaData dmd = new DumpMetaData(loadPath, conf);
 
+        if (!dmd.isVersionCompatible()) {
+          throw new SemanticException("Dump version: " + dmd.getHiveVersion() +
+              ". Versions older than 3 are not supported.");
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java:364]: Dont hardcode 3. Use the constant.

File: ql/src/test/org/apache/hadoop/hive/ql/parse/repl/load/TestDumpMetaData.java [hunk 4]
```diff
@@ -0,0 +1,81 @@
...
+import org.apache.hadoop.hive.ql.parse.repl.DumpType;
+import org.apache.hadoop.hive.ql.parse.repl.dump.Utils;
+import org.junit.Before;
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.List;
+
+import static org.junit.Assert.*;
```
Comment (aasha) [inline ql/src/test/org/apache/hadoop/hive/ql/parse/repl/load/TestDumpMetaData.java:32]: dont use * in the import

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 5]
```diff
@@ -253,6 +253,7 @@
   public static final String MAPNAME = "Map ";
   public static final String REDUCENAME = "Reducer ";
   public static final String ENSURE_OPERATORS_EXECUTED = "ENSURE_OPERATORS_EXECUTED";
+  public static final int MIN_VERSION_FOR_NEW_DUMP_FORMAT = 3;
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:256]: this should be same as dump

File: ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java [hunk 6]
```diff
@@ -372,16 +382,22 @@ private Path getCurrentLoadPath() throws IOException, SemanticException {
           }
         }
         Path hiveDumpPath = new Path(latestUpdatedStatus.getPath(), ReplUtils.REPL_HIVE_BASE_DIR);
-        if (loadPathBase.getFileSystem(conf).exists(new Path(hiveDumpPath,
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java:375]: why is this check removed?

File: ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java [hunk 7]
```diff
@@ -318,27 +317,38 @@ private void analyzeReplLoad(ASTNode ast) throws SemanticException {
...
-                replScope.getDbName(),
-                dmd.getReplScope(),
-                queryState.getLineageState(), evDump, dmd.getEventTo(), dmd.getDumpExecutionId(),
-            initMetricCollection(!evDump, loadPath.toString(), replScope.getDbName(),
-              dmd.getDumpExecutionId()), dmd.isReplScopeModified());
-        rootTasks.add(TaskFactory.get(replLoadWork, conf));
-      } else {
+
+      if (loadPath == null) {
+        return;
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java:322]: Add a log line.

File: ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java [hunk 8]
```diff
@@ -398,11 +398,8 @@ private Path getCurrentLoadPath() throws IOException, SemanticException {
   }
 
   private boolean shouldLoadProceed (Path hiveDumpPath) throws IOException {
-    if (hiveDumpPath != null) { // Is null check necessary?
-      return hiveDumpPath.getFileSystem(conf).exists(new Path(hiveDumpPath, ReplAck.DUMP_ACKNOWLEDGEMENT.toString()))
-          && !hiveDumpPath.getFileSystem(conf).exists(new Path(hiveDumpPath, LOAD_ACKNOWLEDGEMENT.toString()));
-    }
-    return false;
+    return hiveDumpPath.getFileSystem(conf).exists(new Path(hiveDumpPath, ReplAck.DUMP_ACKNOWLEDGEMENT.toString()))
```
Comment (aasha) [inline ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java:401]: better to keep the check. If someone reuses the method, might lead to npe

File: ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java (hunk 6 above)
Comment (pmadhukar-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/parse/ReplicationSemanticAnalyzer.java:375]: Have moved the check to shouldLoadProceed method

**Review by aasha — APPROVED** [review]

**Pull request merged**